nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+      // If robot already should stop, do nothing
+      break;
+    }
+
+    // Process selected polygon
+    if (at == STOP || at == SLOWDOWN) {
+      processStopSlowdown(polygon, collision_points, cmd_vel_in_, robot_action);
+    } else if (at == APPROACH) {
+      processApproach(polygon, collision_points, velocity_, min_collision_time, robot_action);
+    }
+  }
+
+  // In APPROACH model due to sensors noise calculated speed of robot might be instantly changed
+  // causing robot to twitch. To avoid this, when releasing robot after APPROACH,
+  // its speed should be increased gradually smoothing the twitches over time until robot will
+  // reach its desired normal opration.
+  if (robot_action.action_type == DO_NOTHING) {
+    if (robot_action_prev_.action_type == APPROACH) {
+      // Trigger gradually velocity increase
+      release_operation_ = true;
+    }
+
+    if (!release_operation_) {
+      // Continue normal robot operation
+      robot_action.req_vel = cmd_vel_in_;
+    } else {
+      // Release robot from approach: accelerate the robot (without changing speed instantly)
+      // in order to avoid twitching
+      releaseOperation(curr_time, cmd_vel_in_, robot_action);
+    }
+  }
+
+  if (robot_action != robot_action_prev_) {
+    // Report robot action type and changed velocity
+    printAction(robot_action);
+
+    // Publish required robot velocity
+    publishVelocity(robot_action.req_vel);
+
+    // Set previous robot action to avoid duplicating publications
+    robot_action_prev_ = robot_action;
+    // Set previous robot time to calculate speed change when release normal operation
+    prev_time_ = curr_time;
+  }
+
+  publishPolygons();
+}
+
+void CollisionMonitor::processStopSlowdown(
+  const std::shared_ptr<PolygonBase> polygon,
+  const std::vector<Point> & collision_points,
+  const Velocity & velocity,
+  Action & robot_action)
+{
+  if (polygon->getPointsInside(collision_points) > polygon->getMaxPoints()) {
+    if (polygon->getActionType() == STOP) {
+      // Setting up zero velocity for stopping model
+      robot_action.action_type = STOP;
+      robot_action.req_vel.x = 0.0;
+      robot_action.req_vel.y = 0.0;
+      robot_action.req_vel.tw = 0.0;
+    } else {  // SLOWDOWN
+      const Velocity safe_vel = velocity * polygon->getSlowdownRatio();
+      // Check that currently calculated velocity is safer than
+      // chosen for previous shapes one
+      if (robot_action.action_type == DO_NOTHING || safe_vel < robot_action.req_vel) {
+        robot_action.action_type = SLOWDOWN;
+        robot_action.req_vel = safe_vel;
+      }
+    }
+  }
+}
+
+void CollisionMonitor::processApproach(
+    const std::shared_ptr<PolygonBase> polygon,
+    const std::vector<Point> & collision_points,
+    const Velocity & velocity,
+    double & min_collision_time,
+    Action & robot_action)
+{
+  const double collision_time = polygon->getCollisionTime(collision_points, velocity);
+  if (collision_time >= 0.0 &&
```

Comment:
What are each of these 3 conditions doing? If `> 0` then there's a valid non-zero collision time within the maximum time checked for in `time_before_collision_`. We **know** its less than `getTimeBeforeCollision` or else we would have gotten a `-1` from `getCollisionTime`. 

I don't see the value in checking or even storing and receiving `min_collision_time` since you check `safe_vel < robot_action.req_vel` such that if another collision time was shorter than this one, the robot action would already be lower and this would not be utilized

Reply:
> If `> 0` then there's a valid non-zero collision time within the maximum time checked for in time_before_collision_.

If `> 0` collision will be after finite time, if `== 0` collision already happened (more than `max_points` are inside polygon). In first case robot will be slowed down for approaching the obstacle. In second case, `polygon->getSafeVelocity()` will return zero-velocity, which will stop robot. In both cases logic used in the code is the same. Therefore it is written `>=0` condition.

> We know its less than getTimeBeforeCollision or else we would have gotten a -1 from getCollisionTime.

This is made just for the case. Agree, that it is extra check and could be removed

> I don't see the value in checking or even storing and receiving min_collision_time since you check safe_vel < robot_action.req_vel such that if another collision time was shorter than this one, the robot action would already be lower and this would not be utilized

These are not the same conditions: first `collision_time < min_collision_time` shown that we will select minimal time through other polygons with `APPROACH` model. Second condition `safe_vel < robot_action.req_vel` ensures that we selecting safest velocity through all cases. They are similar but still different conditions. Velocities comparison is made by speed value, while time before collision might take into account robot's rotation.

In the corner case, when robot is spinning in place, `collision_time < min_collision_time` might be a necessary check: `safe_vel < robot_action.req_vel` will return `false` while 2 `APPROACH` polygons might give different times before collision, as depicted below:
![checks_optimization](https://user-images.githubusercontent.com/60094858/172903451-d092485e-a094-49d7-91af-e153c9876f33.png)

Reply:
Since, after dynamically changing footprint was introduced, we have only one robot footprint. Hence, these extra conditions were removed and now looks as follows: https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R526. Only one `>= 0` check is remained. The velocity if being checked later in `safe_vel < robot_action.req_vel` condition, as you've mentioned.